battleground_viewer/src/construct_render/render.rs

```diff
@@ -0,0 +1,408 @@
+        }
+    }
+
+    fn meshes_for_pass(
+        &self,
+        pass: RenderPass,
+    ) -> Option<impl Iterator<Item = &Gm<InstancedMesh, M>>> {
+        if (self.participates_in_pass)(pass) {
+            Some(self.meshes.iter())
+        } else {
+            None
+        }
+    }
+
+    pub fn add_primitive(
+        &mut self,
+        batch_hints: BatchProperties,
+        primitive: Primitive,
+        transform: Mat4,
+        color: Color,
+    ) {
+        let key = PrimitiveBatchKey::new(primitive, batch_hints);
+        let batch = &mut self
+            .batches
+            .entry(key.to_batch_key())
+            .or_insert_with(|| PrimitiveBatch {
+                key,
+                transforms: Default::default(),
+                colors: Default::default(),
+            });
+
+        batch.transforms.push(match primitive {
+            Primitive::Line(l) => {
+                use battleground_construct::util::cgmath::ToHomogenous;
+                use battleground_construct::util::cgmath::ToTranslation;
+                let p0_original = vec3(l.p0.0, l.p0.1, l.p0.2);
+                let p1_original = vec3(l.p1.0, l.p1.1, l.p1.2);
+                let p0 = (transform * p0_original.to_h()).to_translation();
+                let p1 = (transform * p1_original.to_h()).to_translation();
+                let rotation = Quat::from_arc(vec3(1.0, 0.0, 0.0), (p1 - p0).normalize(), None);
+                let scale = Mat4::from_nonuniform_scale(
+                    (p0 - p1).magnitude(),
+                    l.width / 2.0,
+                    l.width / 2.0,
+                );
+                Mat4::from_translation(p0) * <_ as Into<Mat4>>::into(rotation) * scale
+            }
+            _ => transform,
+        });
+        batch.colors.push(color);
+    }
+}
+
+impl<M: Material + BatchMaterial> RenderableGeometry for PrimitiveGeometry<M> {
+    fn objects(&self, pass: RenderPass) -> Option<Vec<&dyn Object>> {
+        self.meshes_for_pass(pass)
+            .map(|xs| xs.map(|x| x as &dyn Object).collect::<_>())
+    }
+
+    fn geometries(&self, pass: RenderPass) -> Option<Vec<&InstancedMesh>> {
+        self.meshes_for_pass(pass)
+            .map(|xs| xs.map(|x| &x.geometry).collect::<_>())
+    }
+
+    fn prepare_frame(&mut self) {
+        self.batches.clear();
+        self.meshes.clear();
+    }
+
+    fn finish_frame(&mut self, context: &Context) {
+        for batch in self.batches.values() {
+            let cpu_mesh = primitive_to_mesh(&batch.key.primitive);
+            let material = M::new_for_batch(context, batch.key.properties);
+            let instances = three_d::renderer::geometry::Instances {
+                transformations: batch.transforms.clone(),
+                colors: Some(batch.colors.clone()),
+                ..Default::default()
+            };
+            let gm = Gm::new(InstancedMesh::new(context, &instances, &cpu_mesh), material);
+            self.meshes.push(gm);
+        }
```

Comment:
I feel `finish_frame()` and `prepare_frame()` are flipped around in my mind... It seems this logic definitely has to happen before the frame is rendered? And as such cannot be at the finalization (finish) of the frame? Should we instead assign the meshes in prepare, and in finish clean them up? Or should both actions just go into prepare_frame?

Reply:
The names are misleading since `ConstructRender` does not actually do any rendering 😬

The idea is that `prepare_frame` does whatever it needs to do to prepare for a new frame, and `finish_frame` wraps up bookkeeping and makes sure that the scene is in a renderable state. I'll rename them to `prepare_scene` and `finish_scene`, which will hopefully help place them in the larger context.